databuilder/validate_dummy_data.py

```diff
@@ -1,151 +1,159 @@
-    if dataset_file.suffixes != dummy_data_file.suffixes:
-        expected_extension = "".join(dataset_file.suffixes)
-        msg = f"Expected dummy data file with extension {expected_extension}; got {dummy_data_file.name}"
-        raise DummyDataValidationError(msg)
 
+def validate_csv_against_spec(csv_file, column_specs):
+    reader = csv.DictReader(csv_file)
+    validate_headers(reader.fieldnames, list(column_specs.keys()))
+    empty_columns = set(reader.fieldnames)
 
-def read_into_dataframe(path):
-    """Read data from path into a Pandas DataFrame."""
-    try:
-        suffixes = ".".join([suffix.strip(".") for suffix in path.suffixes])
-        assert suffixes in SUPPORTED_FILE_FORMATS
-        return pd.read_csv(path)
-    except FileNotFoundError:
-        raise DummyDataValidationError(f"Dummy data file not found: {path}")
-
-
-def validate_expected_columns(df, columns_definitions):
-    """Raise DummyDataValidationError if dataframe does not have expected columns."""
-    expected_columns = set(columns_definitions.keys())
-    extra_columns = set(df.columns) - expected_columns
-    if extra_columns:
-        raise DummyDataValidationError(
-            f"Unexpected columns in dummy data: {', '.join(extra_columns)}"
+    for row_n, row in enumerate(reader, start=1):
+        if None in row:
+            raise ValidationError(f"Too many columns on row {row_n}")
+
+        for name, spec in column_specs.items():
+            value = row[name]
+            if value:
+                empty_columns.discard(name)
+            elif value is None:
+                raise ValidationError(f"Too few columns on row {row_n}")
+
+            try:
+                validate_str_against_spec(value, spec)
+            except ValidationError as e:
+                message = f"row {row_n}, column {name!r}: {e}"
+                raise ValidationError(message) from e
+
+    # In the context of dummy data I think that an all-NULL column should be considered
+    # an error, though it might not be strictly speaking invalid
+    if empty_columns:
+        raise ValidationError(f"Columns are empty: {', '.join(empty_columns)}")
+
+
+def validate_headers(headers, expected_headers):
+    headers_set = set(headers)
+    expected_set = set(expected_headers)
+    if headers_set != expected_set:
+        errors = []
+        missing = expected_set - headers_set
+        if missing:
+            errors.append(f"Missing columns: {', '.join(missing)}")
+        extra = headers_set - expected_set
+        if extra:
+            errors.append(f"Unexpected columns: {', '.join(extra)}")
+        raise ValidationError("\n".join(errors))
+    elif headers != expected_headers:
+        # We could be more relaxed about things here, but I think it's worth insisting
+        # that columns be in the same order. We've seen analysis code before which is
+        # sensitive to column ordering.
```

Comment:
The variables in a dataset definition go through a lot of contortions to get here as a list of expected headers. Are you sure that the order is stable?

Reply:
As long as `get_column_definitions()` gives a consistent order then, yes, it is stable. (Since dicts became explicitly insertion-ordered you generally have to go out of your way _not_ to preserve order.)

Reply:
> As long as get_column_definitions() gives a consistent order

Indeed. At the very least we don't have tests asserting that.